Tutorials/WinForms/Classic/implement-crud-functionality-with-xpo-objects.md

```diff
@@ -0,0 +1,169 @@
+        if (e.Clicks == 2) {
+            e.Handled = true;
+            int customerID = (int)CustomersGridView.GetRowCellValue(e.RowHandle, colOid);
+            using (EditCustomerForm form = new EditCustomerForm(customerID)) {
+                form.ShowDialog(this);
+                Reload();
+                CustomersGridView.FocusedRowHandle = CustomersGridView.LocateByValue("Oid", form.CustomerID.Value);
+            }
+        }
+    }
+    private void Reload() {
+        CustomersBindingSource.DataSource = new XPCollection<Customer>(new Session());
+    }
+    ```
+* Use the Visual Studio **Refactor** tool to rename the `Form1_Load` event handler to `CustomersListForm_Load`. To do this, put the cursor at the method name and click the **Edit > Refactor > Rename** menu item or use **Ctrl+R,Ctrl+R**.   
+* Run the application, open the edit Form, change something, and click the **Save** button to see the result.
+## Handle concurrent changes
+>You can skip this part if you are using [InMemoryDataStore](https://docs.devexpress.com/XPO/DevExpress.Xpo.DB.InMemoryDataStore) or an embedded database.
+### Test with multiple users 
+* Build the project and open the output folder (*bin\Debug*) in the file manager. Locate the application executable file and run two application instances.
+* Double click a record in one window to open the edit dialog, and double click the same record in another window.
+* Type something and save changes in each window. The first application saves changes successfully, but the second fails with the following error message: *"Cannot persist the object. It was modified or deleted (purged) by another application"*.
+### Handle the exception and reload an object
+* Open the `EditCustomerForm` designer.
+* Select the [LayoutControl](https://docs.devexpress.com/WindowsForms/DevExpress.XtraLayout.LayoutControl) component and add the [SimpleButton](https://docs.devexpress.com/WindowsForms/DevExpress.XtraEditors.SimpleButton) control from Toolbox.
+* Adjust buttons' positions and sizes.
+* Rename the `simpleButton1` control to **btnReload** and set the `Text` property to **&Reload**.
+* Double click the `btnReload` control to add the `Click` event handler (you can use the **Properties** window to add it).
+* Select all lines in the `EditCustomerForm_Load` method and click the **Edit > Refactor > Extract Method** menu item or use **Ctrl+R,Ctrl+M**.
+* Change the method name to **Reload** and call this method in the `btnReload_Click` event handler.
+* Modify the `btnSave_Click` method.
+    ```csharp
+    using DevExpress.Xpo.DB.Exceptions;
+    // ...
+    private void btnSave_Click(object sender, EventArgs e) {
+        try {
+            UnitOfWork.CommitChanges();
+            Close();
+        } catch (LockingException) {
+            XtraMessageBox.Show(this, "The record was modified or deleted by another user. Please click the Reload button and try again.", "XPO Tutorial", MessageBoxButtons.OK, MessageBoxIcon.Stop);
```

Comment:
```suggestion
            XtraMessageBox.Show(this, "The record was modified or deleted. Click Reload and try again.", "XPO Tutorial", MessageBoxButtons.OK, MessageBoxIcon.Stop);
```

Reply:
I changed this message in the updated example, too.